src/macaron/repo_finder/commit_finder.py

```diff
@@ -0,0 +1,397 @@
+# Copyright (c) 2023 - 2023, Oracle and/or its affiliates. All rights reserved.
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.
+
+"""This module contains the logic for matching PackageURL versions to repository commits via the tags they contain."""
+import logging
+import re
+from re import Pattern
+
+from git import TagReference
+from packageurl import PackageURL
+from pydriller import Git
+
+logger: logging.Logger = logging.getLogger(__name__)
+
+# An optional named capture group "prefix" that accepts one of the following:
+# - A string of any characters starting with an alphabetic character, ending with one of:
+#   - One alphabetic character and one or more numbers.
+#   - One number and one alphabetic character.
+#   - Two alphabetic characters.
+# - OR
+# - Two alphabetic characters.
+# E.g.
+# - name_prefix
+# - prefix-a444
+# - vm
+# - name-prefix-j5u
+PREFIX = "^(?P<prefix>(?:[a-z].*(?:[a-z][0-9]+|[0-9][a-z]|[a-z]{2}))|[a-z]{2})?"
+
+# An optional named capture group "prefix_sep" that accepts one of:
+# - A 'v', 'r', or 'c' character that is not preceded by a non-alphanumeric character.
+# ('c' is probably a typo as it was found in only one example tag, but accepting any single alphabetic character
+# would also most likely be fine.)
+# - A non-alphanumeric character followed by 'v', 'r', or 'c'.
+# - A non-alphanumeric character.
+# Then optionally ending with one non-alphanumeric character.
+# E.g.
+# - _v-
+# - r_
+# - c
+# - .
+PREFIX_SEPARATOR = "(?P<prefix_sep>(?:(?:(?<![0-9a-z])[vrc])|(?:[^0-9a-z][vrc])|[^0-9a-z])(?:[^0-9a-z])?)?"
+
+# Together, the prefix and prefix separator exist to separate the prefix from version part of a tag, while ensuring that
+# the prefix is free from non-prefix characters (the separator). Note that the prefix is expected to be at least two
+# characters in length to prevent overlap with separators and confusion with versions; the prefix separator is at most
+# three characters; and a negative lookback passes if there are no preceding characters.
+
+# The infix exists between parts of the version string. The most recent design resulted in use of a back reference to
+# ensure non-suffix version parts were separated by the same separator, e.g. 1.2.3 but not 1.2-3. However, one edge
+# case required this to be partially reverted, as it required 1.2-3 to be accepted, while another edge case where
+# additional zeros need to be added after a version to pad its length, e.g. 1.2 becomes 1.2.0.0, still requires it.
+
+# The infix accepts either:
+# - One to three alphabetic characters.
+# - One to three non-alphanumeric characters.
+# Note: The upper limit of three could be reduced to two based on current data.
+INFIX_3 = "([a-z]{1,3}|[^0-9a-z]{1,3})"
+INFIX_1 = f"(?P<sep>{INFIX_3})"  # A named capture group of INFIX_3.
+INFIX_2 = "(?P=sep)"  # A back reference to INFIX_1.
```

Comment:
From your implementation, I think these should not be called "infix", because you also have the notion of "prefix" and "suffix" in your tag-matching implementation, and it sounds like "infix" should be "anything and everything between the prefix and the suffix", while in reality these "INFIX_x" patterns here are for "separators between numeric parts in the actual infix of the version string".

If I understand it right, given this version string `foobar-1-2-3-baz`, these infix patterns will match the `-` between `1` and `2`, and `2` and `3`. I think `INFIX_SEP` or something along the line would be a better name to avoid confusion.

I also want to see some examples of these cases in your comments. Similarly to my previous remark, it would be nice if you could also include the full version strings, and then point out which parts are matched as "infix separators".

Reply:
A definition of infix:

> An infix is an affix inserted inside a word stem.

It goes onto say other things but given we are dealing with the creation of regex strings rather than words. I think any addition to the middle parts of the regex would count.

The infix pattern can also match alphabetic characters if they are between the version numbers, e.g. `r1rc2` -> `rc` is an infix.

Given the terminology here, it's probably more accurate to say that the version numbers are the infix separators.

Reply:
For me it's not the word infix by itself that confuses, but the fact that it appears alongside "prefix" and "suffix".
Something like "version_number_separator" is probably better IMO.

Reply:
I have to disagree. The words "prefix", "suffix", and "infix" are all being used to describe parts of the regex pattern that are being added based on where they are being added, not necessarily on how they relate to the version. The base pattern, or word stem from the definition, is simply the parts from the version, e.g. `12A3`.  All "affixes" are modifications to the base pattern that allow the version to be mapped to a tag.

Reply:
Alright. I think as long as you leave some examples in the comments to clarify what these regexes should match, I think it is all good for me. Thanks, Ben. 👍🏼